### PR TITLE
Set ess-insert-assign in the local map

### LINKED ISSUE #659: ess smart assing key no longer works
I have this in my .emacs file. It used to work fine but no longer works. Pressing ";" just inserts ";".

(require 'ess-site)
(require 'ess-rutils)
....
;; smart assign
(setq ess-smart-S-assign-key ";")  ;if ";" needed, press ";" key twice
(ess-toggle-S-assign nil)
(ess-toggle-underscore nil) ;leave underscore key alone!

Thanks.

### PR DESCRIPTION
**Pull request by jabranham**
Closes #659

### CODE REVIEW

**Comment by lionel-** [comment]
There is a small problem. The new binding is registered but the old one is still in effect. People will keep being puzzled as to why `"_"` is still inserting assignment arrows.

I just fixed it locally (I think). I'll push here soon with some tests.

**Comment by lionel-** [comment]
@jabranham Can you look over the new commits please? Should we remove the non-literate smart assign test now there is a more complete set of tests?

**Comment by lionel-** [comment]
One side effect of this change is that it's no longer possible to assign it manually:

```lisp
(define-key ess-mode-map ";" #'ess-insert-assign)
```

Since `ess-smart-S-assign-key` and the bound key have diverged, it looks like the user has redefined the variable so we self-insert. Thus the only way to change the keybinding is to change `ess-smart-S-assign-key`.

I can live with this but I wonder if we wouldn't be better off deprecating `ess-smart-S-assign-key` and teaching our users to bind/unbind the key like any other Emacs command (sorry if this was already suggested, I haven't followed the discussions closely). If we keep the current behaviour we should document the new restriction.

**Comment by lionel-** [comment]
I have split `ess-insert-assign` and `ess-insert-assign--smart-key`. This way we get the best of both worlds:

* `ess-insert-assign` is bindable just like any other key
* `ess-smart-S-assign-key` is configurable effectfully

**Comment by lionel-** [comment]
I don't think many people if any will have binded `"_"` to `self-insert-command`. I think we should merge this branch and then plan for one of two things:

* Keep the smart-key variable, but set it to `nil` by default. If set we override keys in the ESS maps, but at least that's on user action. At this point we can remove some of the code implemented here (the sentinel). People will have to add `(setq ess-smart-S-assign-key "_")` in their config.

* Deprecate the smart key variable and ask users to bind the key themselves.

**Comment by jabranham** [comment]
> I don't think many people if any will have binded "_" to self-insert-command 

This is what I did for a while. I don't think it's OK for ESS to overwrite /any/ changes users have made to ess-mode-map (or any variable after it's defined). That's super un-Emacs-y behavior.

**Comment by lionel-** [comment]
I agree this is not ideal and should be on user action like in the two plans I suggested.

We can also check `"_"` is binded to the smart command or the sentinel, and otherwise set the smart key variable to `nil` if it is equal to `"_"`. I.e. if there's a clash, let the keymap binding prevail.

In general we all have spent too much time on this thing. We should break compatibility and opt for a standard way of setting the key. Also I think `_` should self-insert by default. AFAIK ggplot2 is the most popular package (not counting Rcpp which is not user facing) and it uses snake_case. Using _ for assignment has become an anachronism.

**Comment by jabranham** [comment]
> In general we all have spent too much time on this thing. We should break compatibility and opt for a standard way of setting the key. Also I think _ should self-insert by default. AFAIK ggplot2 is the most popular package (not counting Rcpp which is not user facing) and it uses snake_case. Using _ for assignment has become an anachronism.

This I agree with. If we do this for 18.09 then we can close this PR and just set the smart assign variable to nil, right?

**Comment by lionel-** [comment]
If we opt for the first course of action I suggested in https://github.com/emacs-ess/ESS/pull/660#issuecomment-418758877 this PR is still relevant.

**Comment by jabranham** [comment]
Ah, for changing the binding if the variable is changed to another string during an Emacs session, right.

Seems like it would be easier/cleaner (both for ESS developers and users) to just deprecate the assignment variable and let people assign it manually, that way they could choose what key it gets bound to in the same way that all other keybindings are set up. 

So perhaps the plan could be:

- Set the default to `nil`, mark it as obsolete, and merge this PR. Mention in the news file that it's deprecated and people who want to continue using it should add the binding to their config file themselves.

- Shortly after 18.09, remove the variable and all the infrastructure supporting assignment this way

WDYT?

**Comment by lionel-** [comment]
@jabranham With the last commit user-defined "_" bindings are no longer overwritten.

@vspinu @mmaechler Can you chime in on the discussion above please? This is about one or both of the following:

1. Deprecate `"_"` as default value for `ess-smart-S-assign-key`. Because of this default we need non-conventional code to make everything work smoothly (cf changes in this PR). And this default gets in the way of using packages like ggplot2.

2. Deprecate `ess-smart-S-assign-key` altogether and ask users to define the assign key the normal way.

**Comment by vspinu** [comment]
This PR is way to complex for 18.09. I am really struggling to understand what's going on in there.  Regarding the obsoletion, we will break smart-underscore defaults in 19.04 anyhow and we should not break/obsolete things for 18.09. As #659 suggests, we do have a minor breakage, but if we don't have a simple solution we probably should leave things as they are.

Blindly assigning to local map indeed won't work directly. I myself have only this `(define-key ess-mode-map [?_] nil)` and I don't think it's uncommon. Setting a key to `nil` is a standard way of "unbinding" keys in emacs world. 

One solution would be to check if the smart key is already assigned to something other than `ess-smart-assign` and don't touch it in that case. So for 18.09 I would rather keep the @jabranham's original commit with this extra check added, or just leave it as it is. 

For 19.04 we will either remove ess-smart-assign-key or set it to nil. The latter would mean facing same problems as we do now. So it probably should be the former.

**Comment by lionel-** [comment]
> Blindly assigning to local map indeed won't work directly. I myself have only this (define-key ess-mode-map [?_] nil) and I don't think it's uncommon. Setting a key to nil is a standard way of "unbinding" keys in emacs world.

This all works correctly in this PR. You can set `"_"` to `nil` or a different command and it won't be overridden by the smart key. I think this PR provides the best interaction between the smart key variable and normal key bindings. And it is amply tested by unit tests.

**Comment by vspinu** [comment]
>  I think this PR provides the best interaction between the smart key variable and normal key bindings.

So we don't need to obsolete anything or change the defaults? This is not what I understood from last @jabranham's comment. 

The problem with this underscore business that none of us is testing it in our setup, so we need at least two pairs of eyes to check the logic before we merge any changes. @jabranham are you fine with the changes?

I admit that the only problem that I have with the PR is that I don't understand it :/ I will try to find some time to figure it out today or tomorrow.

**Comment by lionel-** [comment]
> So we don't need to obsolete anything or change the defaults?

I would like to change the default to `nil` to remove some of the code implemented here (the sentinel) and so that snake_case APIs can be used normally with default ESS. But other than that I think the state of things here is good enough if we don't want to break compatibility.

Here is how it works:

* **Smart key set to something else than "_".** We interpret it as a `define-key` at mode startup. The old smart key might still be defined if it was changed in session but it will undefine itself at the next use and self-insert.

* **Smart key set to "_" (default).** There is a sentinel defined in the ESS maps, aliased to self-insert command. We only bind the smart key on mode startup if we found the sentinel. Otherwise this means the user has set the key to `nil` or some other commands which we don't override.

**Comment by lionel-** [comment]
Another issue fixed by this PR: Setting the smart key to `nil` and binding `ess-insert-assign` the standard way. This doesn't work on current master because `ess-insert-assign` self-inserts when the smart key isn't defined.

File: lisp/ess-s-lang.el [hunk 1]
```diff
@@ -597,25 +597,36 @@ keypress to repeat it, so if it is bound to \"C-c C-=\" pressing
...
+  "Insert the first element of `ess-assign-list'.
 If the character before point is the first element of
-`ess-assign-list', replace it with the last character typed. If
-`ess-smart-S-assign-key' is nil, do `self-insert-command' using
-ARG as the number of times to insert."
+`ess-assign-list', replace it with DEFAULT."
+  (interactive)
+  (let ((assign (car ess-assign-list)))
+    (if (re-search-backward assign (- (point) (length assign)) t)
+        (replace-match (or default "") t t)
```
[the remark this answers is not in the document]
Comment (lionel-) [reply lisp/ess-s-lang.el]: it seemed cleaner to separate the smart behaviour, but I hadn't realised we had `ess-insert-assign` before. I see that it's an alias for `-S-assign`. We can just merge the functions.

File: lisp/ess-s-lang.el [hunk 2]
```diff
@@ -597,25 +597,36 @@ keypress to repeat it, so if it is bound to \"C-c C-=\" pressing
...
+
+(defun ess-insert-assign--smart-key (arg)
   (interactive "p")
   (if (and ess-smart-S-assign-key
            (string= ess-language "S"))
       (let* ((assign (car ess-assign-list))
              (event (event-basic-type last-input-event))
              (char (ignore-errors (format "%c" event))))
-        (cond ((and char (ess-inside-string-or-comment-p))
+        (cond ((and char (not (string= char ess-smart-S-assign-key)))
```
[the remark this answers is not in the document]
Comment (lionel-) [reply lisp/ess-s-lang.el]: The alternative is to keep the last binding in a variable, then we can unbind it as we bind the new one. However I don't see any problem with the smart key unbinding itself, I even find it more elegant.

**Comment by vspinu** [comment]
I would propose the following simple solution. Make local binding as in Alex's first commit and check if the "standard value" of ess-smart-S-assign-key is bound to `ess-insert-assign` and unbind it if so. This would very close in practical terms to the Lionel's patch but much simpler.

File: lisp/ess.el [hunk 3]
```diff
@@ -111,8 +111,7 @@
     (define-key map "\C-c\C-d"   'ess-doc-map)
     (define-key map "\C-c\C-e"   'ess-extra-map)
     (define-key map "\C-c\C-t"   'ess-dev-map)
-    (when ess-smart-S-assign-key
-      (define-key map ess-smart-S-assign-key 'ess-insert-assign))
+    (define-key map "_"          'ess--smart-assign-sentinel)
```
[the remark this answers is not in the document]
Comment (lionel-) [reply lisp/ess.el]: Not a big deal?

**Comment by lionel-** [comment]
Yeah, I guess it's unlikely the variable will be changed twice in a session and your solution covers the first change. But I would still set the variable to `nil` by default and then we can forget about it forever without extra configuration in our dotfile.

**Comment by vspinu** [comment]
Let's just leave things as they are. We are spending too many brain cells on this tiny issue. All of these will be wiped up after the release when we deprecate the insert key.

**Comment by lionel-** [comment]
I definitely agree we've wasted too much time on this thing. However we can't release the current state in master because this doesn't work:

```lisp
(setq ess-smart-S-assign-key nil)
(define-key ";" ess-mode-map #'ess-insert-assign)
```

Also the unit tests here are relevant.

**Comment by vspinu** [comment]
Indeed. Sorry about that. Just wanted to move forward.

**Pull request closed without merging**